### PR TITLE
Cxx 2496

### PR DESCRIPTION
**Pull request by kkloberdanz**
CXX-2496

Add FLE 2 behavior for CreateCollection() and Collection.Drop()

- Sync the fle2-CreateCollection.yml CSFLE test.
- Add encryptedFieldsMap option to AutoEncryptionOpts
- Add encryptedFields options to CreateCollection and Collection.Drop()
- Update behavior of CreateCollection and Collection.Drop() when the target collection has encryptedFields.

### CODE REVIEW

**Comment by kkloberdanz** [comment]
Not sure what could be causing this warning from the POWER build:
```
[2022/12/14 21:07:19.847] cd /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/build/src/mongocxx/test && /usr/bin/c++  -DBSONCXX_TESTING -DMONGOCXX_TESTING -DMONGO_CXX_DRIVER_COMPILING -I/data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/third_party/catch/include -I/include -I/data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/.. -I/data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/build/src/mongocxx/.. -I/data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/build/src/bsoncxx/third_party/EP_mnmlstc_core-prefix/src/EP_mnmlstc_core/include -I/data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/build/install/include/bsoncxx/v_noabi -I/data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/bsoncxx/.. -I/data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/build/src/bsoncxx/.. -isystem /data/mci/0776864d158095bf494026871536bc9c/mongoc/include/libmongoc-1.0 -isystem /data/mci/0776864d158095bf494026871536bc9c/mongoc/include/libbson-1.0  -Wall -Wextra -Wconversion -Wnarrowing -Wno-expansion-to-defined -pedantic -Werror -Wno-missing-field-initializers -Wno-aligned-new -O3 -DNDEBUG   -Wno-maybe-uninitialized -std=c++11 -o CMakeFiles/test_driver.dir/write_concern.cpp.o -c /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/test/write_concern.cpp
[2022/12/14 21:07:30.829] found libmongoc version 1.23.0-pre
[2022/12/14 21:07:30.829] In file included from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/../bsoncxx/stdx/optional.hpp:22,
[2022/12/14 21:07:30.829]                  from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/../bsoncxx/view_or_value.hpp:19,
[2022/12/14 21:07:30.829]                  from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/../bsoncxx/array/view_or_value.hpp:20,
[2022/12/14 21:07:30.829]                  from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/../bsoncxx/types/bson_value/value.hpp:21,
[2022/12/14 21:07:30.829]                  from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/test/spec/unified_tests/entity.hh:20,
[2022/12/14 21:07:30.829]                  from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/test/spec/unified_tests/assert.hh:17,
[2022/12/14 21:07:30.829]                  from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/test/spec/unified_tests/runner.cpp:21:
[2022/12/14 21:07:30.829] /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/build/src/bsoncxx/third_party/EP_mnmlstc_core-prefix/src/EP_mnmlstc_core/include/core/optional.hpp: In function ‘bool {anonymous}::add_to_map(const bsoncxx::v_noabi::array::element&)’:
[2022/12/14 21:07:30.829] /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/build/src/bsoncxx/third_party/EP_mnmlstc_core-prefix/src/EP_mnmlstc_core/include/core/optional.hpp:108:5: error: ‘*((void*)& __tmp +6)’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
[2022/12/14 21:07:30.829]      ::new(::std::addressof(this->val)) value_type {
[2022/12/14 21:07:30.829]      ^~
[2022/12/14 21:07:30.829] In file included from /usr/include/c++/8/bits/nested_exception.h:40,
[2022/12/14 21:07:30.829]                  from /usr/include/c++/8/exception:144,
[2022/12/14 21:07:30.830]                  from /usr/include/c++/8/ios:39,
[2022/12/14 21:07:30.830]                  from /usr/include/c++/8/istream:38,
[2022/12/14 21:07:30.830]                  from /usr/include/c++/8/fstream:38,
[2022/12/14 21:07:30.830]                  from /data/mci/0776864d158095bf494026871536bc9c/mongo-cxx-driver/src/mongocxx/test/spec/unified_tests/runner.cpp:15:
[2022/12/14 21:07:30.830] /usr/include/c++/8/bits/move.h:193:11: note: ‘*((void*)& __tmp +6)’ was declared here
[2022/12/14 21:07:30.830]        _Tp __tmp = _GLIBCXX_MOVE(__a);
[2022/12/14 21:07:30.830]            ^~~~~
[2022/12/14 21:07:32.860] cc1plus: all warnings being treated as errors
```

**Review by rcsanchez97 — APPROVED** [review]
LGTM (though C++ is far from my strong suit, so take this with a grain of salt).

**Comment by rcsanchez97** [comment]
> Not sure what could be causing this warning from the POWER build:
> 
I poked around at this for a bit and it isn't clear to me why this is an error on just this particular platform.  If you'd like it, I can try to help dig deeper.  Let me know.

**Comment by kkloberdanz** [comment]
> > Not sure what could be causing this warning from the POWER build:
> 
> I poked around at this for a bit and it isn't clear to me why this is an error on just this particular platform. If you'd like it, I can try to help dig deeper. Let me know.

Thanks! I appreciate it! I'm not quite sure what's causing it.

File: CMakeLists.txt [hunk 1]
```diff
@@ -40,6 +40,10 @@ else()
   message(WARNING "Unknown compiler... recklessly proceeding without a version check")
 endif()
 
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "ppc64le")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-maybe-uninitialized")
+endif()
```
Comment (kkloberdanz) [inline CMakeLists.txt:45]: I'm open to suggestions on where is the best place to put this check.

**Comment by rcsanchez97** [comment]
Doesn't `-Wno-maybe-uninitialized` disable the warning entirely?  I
would think that `-Wno-error=maybe-uninitialized` is more useful as it
still allows the warnings to be produced but without failing the build.

As far as location, where you have it works perfectly well, or you could
move it down a bit to just before the line with
`set (CMAKE_SKIP_BUILD_RPATH false)`.  That would be my preferred
location as this is new check you are adding is definitely the most
obscure of the checks.

**Review by kevinAlbs — APPROVED** [review]
Nicely done. LGTM

**Review by vector-of-bool — APPROVED** [review]
LGTM

**Pull request merged**